torch/_inductor/simple_fsdp/fsdp_reorder.py

```diff
@@ -0,0 +1,181 @@
+# mypy: allow-untyped-defs
+# pyre-strict
+import math
+from enum import IntEnum
+from collections import defaultdict
+from typing import Dict, List, Optional, Set, Tuple
+
+import torch
+import torch.distributed as dist
+from .. import dependencies, ir, scheduler
+from ..virtualized import V
+
+from .utils import NodeType, compute_bucket_users, get_node_type
+
+
+def reorder_all_gather(
+    snodes: List["scheduler.BaseSchedulerNode"],
+    all_gather_before_last_wait: Optional[bool] = True,
+) -> List["scheduler.BaseSchedulerNode"]:
+    """
+    Reorder All Gather and Wait in the forward/backward pass;
+    1. all_gather_before_last_wait set to True: all_gather_i is reordered before wait_i-1
+    2. all_gather_before_last_wait set to False: all_gather_i is reordered after wait_i-1
+    """
+    result_list: List[scheduler.BaseSchedulerNode] = []
+    all_gather_list: List[scheduler.BaseSchedulerNode] = []
+    node_to_type: Dict[scheduler.BaseSchedulerNode, int] = {}
+    inverse_users, node_users = compute_bucket_users(snodes)
+    
+    for node in snodes:
+        node_to_type[node] = get_node_type(node)
+    
+    snodes.reverse()
+    for idx, node in enumerate(snodes):
+        node_type = node_to_type[node]
+        if node_type in [NodeType.REDUCE_SCATTER, NodeType.COMPUTE, NodeType.RS_WAIT]:
+            # we do not reorder reduce scatter and compute node
+            if node not in result_list and node not in all_gather_list:
+                result_list.append(node)
+        elif node_type == NodeType.ALL_GATHER:
+            # gather i-th all gather node and its dependencies
+            all_gather_list.append(node)
+            inverse_user = list(inverse_users[node])
+            if (
+                len(inverse_user) > 0
+                and not node_to_type[inverse_user[0]] == NodeType.ALL_GATHER
+                and inverse_user not in all_gather_list
+            ):
+                all_gather_list.extend(inverse_user)
+        elif node_type == NodeType.AG_WAIT:
+            if (
+                (
+                    (
+                        node_to_type[snodes[idx + 1]] == NodeType.AG_WAIT
+                        and node_to_type[snodes[idx + 2]] == NodeType.ALL_GATHER
+                    )
+                    or (
+                        node_to_type[snodes[idx + 1]] == NodeType.ALL_GATHER
+                        and node_to_type[snodes[idx - 1]] != NodeType.AG_WAIT
+                    )
+                )
+                and not all_gather_before_last_wait
+                and len(all_gather_list) > 0
+            ):
```

Comment:
the code is difficult to understand... we probably should have more categories in `NodeType`, e.g. copy-in, copy-out